### PR TITLE
Resumable download

### PR DESCRIPTION
**Pull request by YanisOUALAN**
### For Android and iOS : 
This pull request introduces a resumable download feature to the updater, addressing the issue reported in #255. 

The new functionality ensures that if the application is closed, receives a SIGKILL signal, or crashes during an update, the update process will resume from the last completed point the next time the application is launched. 

This enhancement aims to provide a more robust update experience by mitigating the impact of disruptions during the update process while reducing bandwidth usage.

You can see it in action here : https://www.youtube.com/watch?v=Y-MFS3PfmCc

Let me know if any question !

/claim #255

### CODE REVIEW

**Comment by riderx** [comment]
Thanks for the work, this should be done in Android and IOS, can you work on the ios side?

**Comment by YanisOUALAN** [comment]
> Thanks for the work, this should be done in Android and IOS, can you work on the ios side?

Hi, I'm currently working on it

**Comment by YanisOUALAN** [comment]
Just added resumable download for iOS in 37efd56

File: android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java [hunk 1]
```diff
@@ -54,84 +45,111 @@ protected void onHandleIntent(Intent intent) {
...
+
     }
   }
 
   private void notifyDownload(String id, int percent) {
     Intent intent = new Intent(PERCENTDOWNLOAD);
     intent.setPackage(getPackageName());
     intent.putExtra(ID, id);
-    intent.putExtra(PERCENT, percent);
+    intent.putExtra("percent", percent);
```
Comment (riderx) [inline android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java]: Why stop using the static?

**Comment by riderx** [comment]
Thanks for the work!
@neo773 and @ayewo if you are around to help on the review, I can support it :)

File: android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java (hunk 1 above)
Comment (yanisoln) [reply android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java]: Just a mistake, fixed it back in a16b2e0

**Comment by ayewo** [comment]
>  ...  if you are around to help on the review, I can support both :) As it's critical for Capgo

Hey @riderx 👋!

I don't see a sibling PR from the author for the [`Cap-go/capgo/`](https://github.com/Cap-go/capgo/pulls) repo so I need to ask clarifying questions since I've been away from the code base for a while now. 

Regarding my earlier [comment](https://github.com/Cap-go/capacitor-updater/issues/255#issuecomment-1731213184):
1. It seems a separate endpoint is no longer needed now that the expiration time for a download URL from S3 is now [1 week](https://github.com/Cap-go/capgo/blob/9942a96440c70c7d85d23a863eb0d35fbfb664ab/supabase/functions/_backend/utils/downloadUrl.ts#L6) instead of 2mins?

2. Is the service still using Cloudflare caching?

**Comment by riderx** [comment]
@ayewo you are correct we now extended the validity of the zip as it was required for other cache test.
We do use cloudflare r2, but not the cache i think
In the future i think we will even remove the singing to be able to benefit from CND of cloudflare r2 with custom domain name.

**Comment by ayewo** [comment]
Hi @YanisOUALAN,

can you please provide step-by-step instructions that can be used to verify your implementation works as desired?

Including the demo app you used for testing would be nice too.

Thanks!

**Comment by YanisOUALAN** [comment]
> Hi @YanisOUALAN,
> 
> can you please provide step-by-step instructions that can be used to verify your implementation works as desired?
> 
> Including the demo app you used for testing would be nice too.
> 
> Thanks!

Hey, sure,

### To setup the demo app:

1. Download the app [here](https://drive.google.com/file/d/1Ck9DJGGitSzduvJp1kgrqGKZOXukBUcF/view?usp=sharing)
2. Extract it and all the stuff
3. Install capacitor with `npm init @capacitor/app`

### Install my version of capgo :
4. Run in the demo app folder the command `git clone https://github.com/YanisOUALAN/capacitor-updater.git`
5. Move in the repo with `cd capacitor-updater`
6. Move to the `resumable_download` branch with `git checkout resumable_download`
7. Run `npm install` and `npm run build`
8. Create a symbolic link with `npm link`
9. Go back in the demo app folder with `cd ../` and add reference to the symbolic link with `npm link @capgo/capacitor-updater` then run `npm run build`

### Create a bundle for triggering the update :
10. Give to the app a unique package name in `capacitor.config.ts`
11. Init the capgo setup with the command on [your capgo cloud dashboard](https://web.capgo.app/app/home) or setup your self hosted solution. **When it ask if you want to add` Automatic Add "CapacitorUpdater.notifyAppReady()" code and import` say no as it's already added. Same thing when it ask to `Automatic Install "@capgo/capacitor-updater" dependency`**
12. Add your platform with `npx cap add android` and/or `npx cap add ios` 
13. Run `npm run build` and `npx cap sync`
14. If you're using capgo cloud, setup the production channel as the default channel
15. Modify in demo app package.json the attribute `"version": "1.1.x",` with `"version": "1.1.x+1",` you can also modify the texts in `src/App.js` file to make the change more clear when the app is updated
16. Run the command `npm run build`
17. Upload your bundle to trigger an update the next time the app is launched, if you're using capgo cloud, use the command `npx @capgo/cli@latest bundle upload --channel=production`
18. If you're using capgo cloud, setup the `production` channel as the default channel
19. Test your app on Android Studio or xCode with `npx cap open android` or/and `npx cap open ios`


### For testing (iOS and Android) : 

The termination (SIGKILL) scenario :

1. Build & launch the demo app
2. Wait for the download to start downloading
3. Before the download has reached 100%, kill the app (as i did in both videos i showed before)
4. Restart the app
5. Ensure that the download is not restarting from 0 but from where it was before the app get killed

The disconnected scenario :

1. Build & launch the demo app
2. Wait for the download to start downloading
3. Before the download has reached 100%, disconnect your emulator/your host from internet
4. Close the app
5. Bring your connection back
6. Restart the app
7. Ensure that the download is not restarting from 0 but from where it was before the app get disconnected

The app closed scenario :

1. Build & launch the demo app
2. Wait for the download to start downloading
3. Before the download has reached 100%, open the multitasking menu on your device (swipe the navigation bar up) and close the app (by sliding it up)
4. Restart the app
5. Ensure that the download is not restarting from 0 but from where it was before the app get closed


I suggest to add these lines to facilitate the debugging in `node_modules/@capgo/capacitor-updater/android/src/main/java/ee/forgr/capacitor_updater/DownloadServices.java` capgo file just after `int percent = calcTotalPercent(downloadedBytes, contentLength);` on line 118 :
```
System.out.print("Download progress : " + percent);
```
And on `node_modules/@capgo/capacitor-updater/ios/Plugin/CapacitorUpdater.swift`, just after `let percent = Int((Double(totalReceivedBytes) / Double(targetSize ?? 1)) * 100.0)` on line 628 :
```
print("Downloading : \(percent)%")
```
In order to show the download progress

I also recommend using a network throttler like [NetLimiter for Windows](https://www.netlimiter.com/) or [Network Link Conditioner for MacOS](https://apple.stackexchange.com/questions/24066/how-to-simulate-slow-internet-connections-on-the-mac) so you can slow down the download progress artificially, otherwise it will happen too quickly before you can try to stop it.

If the download happened and your app got updated and you want to test the downloading process again, just go to package.json in the demo app folder, and change `"version": "1.1.x",` to `"version": "1.1.x+1",` run `npm run build` and upload a new bundle with `npx @capgo/cli@latest bundle upload --channel=production` note that it will just trigger the download process again, the react view and the associated text won't change, unless you change them.

Let me know if you have any question !

**Comment by riderx** [comment]
Hey @ayewo do you have time to test it ?

**Comment by ayewo** [comment]
Hey @riderx 

Sorry for the delay. If it's not too late, I should have some availability to review it either tomorrow or Saturday, God willing.

**Comment by riderx** [comment]
Yes it's good for us ! i will make it good pay as well as it's important

File: android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java [hunk 2]
```diff
@@ -54,52 +44,104 @@ protected void onHandleIntent(Intent intent) {
...
+    } catch (IOException e) {
       e.printStackTrace();
-      publishResults(
-        "",
-        id,
-        version,
-        checksum,
-        sessionKey,
-        e.getLocalizedMessage()
-      );
```
Comment (ayewo) [inline android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java]: Why did you remove this code?

Comment (yanisoln) [reply android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java]: Yup mb, forgot to bring back out of memory handler while rewriting the code logic, brought it back in [33d9300](https://github.com/Cap-go/capacitor-updater/pull/400/commits/33d9300db909c8fa3bf8158c61251b89b581e396)

**Comment by YanisOUALAN** [comment]
Have you guys managed to test my implementation ? Let me know if you have any issues 😉 !

File: ios/Plugin/CapacitorUpdater.swift [hunk 3]
```diff
@@ -556,74 +557,173 @@ extension CustomError: LocalizedError {
         print("\(self.TAG) Current bundle set to: \((bundle ).isEmpty ? BundleInfo.ID_BUILTIN : bundle)")
     }
 
-    public func download(url: URL, version: String, sessionKey: String) throws -> BundleInfo {
-        let semaphore: DispatchSemaphore = DispatchSemaphore(value: 0)
-        let id: String = self.randomString(length: 10)
-        var checksum: String = ""
+    //Do a GET request on the url in order to extract the Content-Length header, ask only for the 10 first bytes in order to not download the full content.
```
Comment (WcaleNieWolny) [inline ios/Plugin/CapacitorUpdater.swift]: I believe that you could use a "head" request in order to get the content-length. As far as I know we do that [here](https://github.com/Cap-go/capgo/blob/5517026d4640659bd9020583225e01b4634d0047/supabase/functions/_backend/utils/s3.ts#L124)

Comment (yanisoln) [reply ios/Plugin/CapacitorUpdater.swift]: Apparently, the server doesn't support HEAD requests; at least, when I tried to do so, I got a 403 response from the server, so I had to work around with GET and the Range header.
capgo head request

Comment (riderx) [reply ios/Plugin/CapacitorUpdater.swift]: We do head request just here: https://github.com/Cap-go/capgo/blob/ebde1312a51feb6288ea5be750df71f3399cab46/supabase/functions/_backend/utils/s3.ts#L116
To get the size.
Just after getSignedUrl who is the function used to respond to the update (what you are getting), so it should work

Comment (yanisoln) [reply ios/Plugin/CapacitorUpdater.swift]: Please let me know if I'm missing anything. I attempted to use a `HEAD` request to retrieve the `Content-Length` header, but I kept receiving a `403` error, which suggests that the download URL may not support `HEAD` requests. To address this issue, in commit 90a09dd, I modified my `GET` request logic to include `Range: bytes=0-0`. This adjustment ensures that my `GET` request is basically the same as a `HEAD` request, effectively retrieving only the first byte and **providing the necessary headers**.
403capgo

**Comment by riderx** [comment]
Thanks a lot @YanisOUALAN we are almost ready to merge, now we have a little conflict, if you can solve it, we merge

File: android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java [hunk 4]
```diff
@@ -1,19 +1,12 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at https://mozilla.org/MPL/2.0/.
```
Comment (riderx) [inline android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java:4]: Please add back the header it's a requirement of the licence

Comment (yanisoln) [reply android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java:4]: added back in 0fa27d2 !

**Comment by YanisOUALAN** [comment]
> Thanks a lot @YanisOUALAN we are almost ready to merge, now we have a little conflict, if you can solve it, we merge

Done ! ✅

**Comment by riderx** [comment]
Thanks ! We did review one more time and we have a last issue.
Our current code style try to follow the same name/structure for each platform. Currently, both do differ a lot, do you have a way to make them closer ?

**Comment by WcaleNieWolny** [comment]
Hello, there is an critical issue with the android implementation

https://github.com/user-attachments/assets/5ae295f4-5b10-453a-b2fd-4e9921e2f42c

File: android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java [hunk 5]
```diff
@@ -54,39 +50,102 @@ protected void onHandleIntent(Intent intent) {
...
+
+      // Reading progress file (if exist)
+      long downloadedBytes = 0;
+      if (progressFile.exists() && tempFile.exists()) {
+        try (
+          BufferedReader reader = new BufferedReader(
+            new FileReader(progressFile)
+          )
+        ) {
+          downloadedBytes = Long.parseLong(reader.readLine());
```
Comment (WcaleNieWolny) [inline android/src/main/java/ee/forgr/capacitor_updater/DownloadService.java]: ProgressFile can be replaced with `downloadedBytes = tempFile.length();`

File: ios/Plugin/CapacitorUpdater.swift [hunk 6]
```diff
@@ -555,75 +556,202 @@ extension CustomError: LocalizedError {
...
-            return (fileURL, [.removePreviousFile, .createIntermediateDirectories])
+            return String(format: "%08X", checksum).lowercased()
+        } catch {
+            print("\(self.TAG) Cannot get checksum: \(filePath.path)", error)
+            return ""
         }
-        let request = AF.download(url, to: destination)
+    }
+    //Do a GET request on the url in order to extract the Content-Length header, ask only for the 10 first bytes in order to not download the full content.
+    func fetchFileSize(url: URL) -> Int? {
```
Comment (WcaleNieWolny) [inline ios/Plugin/CapacitorUpdater.swift]: This function is not required.

You can get the length of the file that you are downloading by doing the following

```swift
        //Opening connection for streaming the bytes
        AF.streamRequest(url, headers: requestHeaders).validate().onHTTPResponse(perform: { response in
            print("magic \(response.headers)")
        }).responseStream { [weak self] streamResponse in
```

And it will return the following:

```
magic Etag: "1869b2563b642e042abdf422cf58884b"
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Amz-Request-Id: 17EB9AD5E895213C
X-Amz-Id-2: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8
X-Content-Type-Options: nosniff
Content-Range: bytes 0-100664/100665
Date: Wed, 14 Aug 2024 13:18:45 GMT
Content-Type: binary/octet-stream
Server: MinIO
Vary: Origin, Accept-Encoding
Last-Modified: Wed, 14 Aug 2024 12:44:58 GMT
Accept-Ranges: bytes
X-Xss-Protection: 1; mode=block
Content-Length: 100665 <--------- IMPORTANT HEADER
```

**Review by WcaleNieWolny — CHANGES_REQUESTED** [review]
Please resolve the critical android issue that I showed on the video (likely also present in IOS). Please also check if it's possible to avoid using `fetchFileSize` in IOS

File: ios/Plugin/CapacitorUpdater.swift [hunk 7]
```diff
@@ -673,20 +677,33 @@ extension CustomError: LocalizedError {
...
+    private func saveDownloadInfo(_ version: String) {
         do {
-            try "\(bytes)".write(to: progressPath, atomically: true, encoding: .utf8)
+            try "\(version)".write(to: updateInfo, atomically: true, encoding: .utf8)
         } catch {
             print("Failed to save progress: \(error)")
         }
     }
-
+    private func getLocalUpdateVersion() -> String { //Return the version that was tried to be downloaded on last download attempt
```
Comment (WcaleNieWolny) [inline ios/Plugin/CapacitorUpdater.swift:809]: Hello, this is a good step forward, but I think that this function could be simplfied

```swift
UserDefaults.standard.string(forKey: "PROGRESSIVE_DOWNLOAD_LAST_VERSION")
 ```
 
 You would set it with
 ```swift
UserDefaults.standard.set("1.0.0", forKey: "PROGRESSIVE_DOWNLOAD_LAST_VERSION")
UserDefaults.standard.synchronize()
```

A similar thing can be done in android

Comment (yanisoln) [reply ios/Plugin/CapacitorUpdater.swift:809]: Hi 👋 ,

The issue with UserDefaults is that it saves data to the disk at regular intervals, not immediately. This means that if the app crashes or is terminated before UserDefaults has a chance to save the data, the app will not recognize any changes made since the last save, and the download will start from 0 again.

The .synchronize method you are proposing [is not recommended to be used by Apple](https://developer.apple.com/documentation/foundation/userdefaults/1414005-synchronize) and is likely to be deprecated in the future, given these concerns, it might not be prudent to rely on this approach.

So it depends if we can neglect this aspect or not, if not, I believe my method might offer a safer alternative !

File: ios/Plugin/CapacitorUpdater.swift (hunk 6 above)
Comment (yanisoln) [reply ios/Plugin/CapacitorUpdater.swift]: Thanks for the help ! I adressed a solution and removed `fetchFileSize` in [fbb44b4](https://github.com/Cap-go/capacitor-updater/pull/400/commits/fbb44b4b41cffb22b41e15db902f4b25c8531e91)

**Comment by YanisOUALAN** [comment]
> Please resolve the critical android issue that I showed on the video (likely also present in IOS). Please also check if it's possible to avoid using `fetchFileSize` in IOS

I resolved the issue for Android in [c98b926](https://github.com/Cap-go/capacitor-updater/pull/400/commits/c98b9264349cb6e5c969bcfbf6624d2dbf6c5f32) and for iOS in [defc590](https://github.com/Cap-go/capacitor-updater/pull/400/commits/defc59000f53111d2a4dab2ca7865b47e2af741b), I also removed `fetchFileSize` and used your approach in [fbb44b4](https://github.com/Cap-go/capacitor-updater/pull/400/commits/fbb44b4b41cffb22b41e15db902f4b25c8531e91), let me know if you find any more issues !

File: ios/Plugin/CapacitorUpdater.swift [hunk 8]
```diff
@@ -555,75 +529,183 @@ extension CustomError: LocalizedError {
         UserDefaults.standard.synchronize()
         print("\(self.TAG) Current bundle set to: \((bundle ).isEmpty ? BundleInfo.ID_BUILTIN : bundle)")
     }
+    private func getChecksum(filePath: URL) -> String {
```
Comment (WcaleNieWolny) [inline ios/Plugin/CapacitorUpdater.swift]: I believe that this function was renamed: https://github.com/Cap-go/capacitor-updater/commit/c309a6f58f706d997ad33f44dba8aab4cad2eb36

File: ios/Plugin/CapacitorUpdater.swift (hunk 7 above)
Comment (WcaleNieWolny) [reply ios/Plugin/CapacitorUpdater.swift:809]: Hi, thank you for brining this to my attention. You are right, the second file is a good idea

**Comment by WcaleNieWolny** [comment]
Hey, I like the changes made to this PR. It looks good to me for now but I will boot up an emulator to make some final checks

File: ios/Plugin/CapacitorUpdater.swift [hunk 9]
```diff
@@ -555,75 +529,183 @@ extension CustomError: LocalizedError {
...
+                 
+             case .stream(let result):
+                 if case .success(let data) = result {
+                     self.tempData.append(data)
+                     
+                     self.savePartialData(startingAt: UInt64(totalReceivedBytes)) // Saving the received data in the package.tmp file
+                     totalReceivedBytes += Int64(data.count)
+                     
+                     let percent = Int((Double(totalReceivedBytes) / Double(targetSize)) * 100.0)
+                     print("Downloading : \(percent)%")
```
Comment (WcaleNieWolny) [inline ios/Plugin/CapacitorUpdater.swift]: I believe that this is a debug log. I like it, but I would like it to consistent between IOS and Android.

You can do the following:
1. Remove it
2. Add a `\(self.TAG)` and add a similar log to android

**Comment by WcaleNieWolny** [comment]
I will be committing some final changes so that this can be merged into main. Great work today 🚀

**Comment by WcaleNieWolny** [comment]
Hello, there is a major problem in IOS. The code is lacking error handling for a failed download. Could you check if it;s possible to add it?

**Comment by WcaleNieWolny** [comment]
Hey, can you please replicate the `stats` (download percent) in android?

**Comment by YanisOUALAN** [comment]
> 2049230

I don't know if that's exactly what you are talking about, but in [2049230](https://github.com/Cap-go/capacitor-updater/pull/400/commits/2049230c2760110fec523db80acbc72412d5359e) i changed the way percentage is calculated by doing the calculation without the calcTotalPercent function

**Comment by riderx** [comment]
@YanisOUALAN calcTotalPercent should still exist as in IOS.
And our download % is not only counting download but all steps, unzip and etc so it's important that the download even of pure download stay between 10 and 70 and the other are still present

**Comment by riderx** [comment]
My bad the download seem still correct just keep the function

**Comment by riderx** [comment]
We try to make the code look the same between platform as much as we can

**Comment by WcaleNieWolny** [comment]
> Our download % is not only counting download but all steps, unzip and etc so it's important that the download even of pure download stay between 10 and 70 and the other are still present

@riderx I don't think that it's the case. I have not observed this behavior for a long while now. Currently `download_xy` does not work at all from what I've seen

I strongly believe that `download_xy` should be for download ONLY and not unzip, etc. 
I believe we should make unzip a `0/1` function (either works or doesn't and we send an error)

**Comment by WcaleNieWolny** [comment]
@riderx we MUST start unifying the code base. This is a great opportunity to fix the download_%.
I was also thinking about backporting [this change](https://github.com/Cap-go/capacitor-updater/pull/429/files#diff-915fe851318ff75f14c8fe44a222ccc19befe3cec76549c7250e00660e9801c3R761-R780) because it's a part of `download` and this PR fully rewrote download in IOS

**Comment by WcaleNieWolny** [comment]
The thing about the download % is that we have NO WAY of knowing that it's going to be a multiple of 10. 
In IOS, we don't control the chunk size so for small bundles + fast internet connections you may get the following:

`percent: 12`
`percent: 53`
`percent 89`
`percent 100`

I think that they are perfectly valid and that we should set them as a `stat`. I would say that we want to send a `stat` when the latest sent stat is a least 10 points different than the last stat that we sent like i did [here](https://github.com/Cap-go/capacitor-updater/pull/400/files#diff-3595c562fbba198d09652f088f74f7368429d9e551a24e04dd86577ea1e03681R666-R669)

**Comment by WcaleNieWolny** [comment]
Alright, I would say that IOS is quite nicely polished. The one thing that is left in IOS is figuring out the `download_%` stat

**Comment by YanisOUALAN** [comment]
> The thing about the download % is that we have NO WAY of knowing that it's going to be a multiple of 10. In IOS, we don't control the chunk size so for small bundles + fast internet connections you may get the following:
> 
> `percent: 12` `percent: 53` `percent 89` `percent 100`
> 
> I think that they are perfectly valid and that we should set them as a `stat`. I would say that we want to send a `stat` when the latest sent stat is a least 10 points different than the last stat that we sent like i did [here](https://github.com/Cap-go/capacitor-updater/pull/400/files#diff-3595c562fbba198d09652f088f74f7368429d9e551a24e04dd86577ea1e03681R666-R669)

As a possible solution, we could ensure every 10% milestone is recorded once a chunk completes. For instance, for a reported percent: 53, we would send :

- `download_0`
- `download_10`
- `download_20`
- `download_30`
- `download_40`
- `download_50`

if tracking precise milestones is crucial.

**Comment by WcaleNieWolny** [comment]
We could, ultimately this decision is up to Martin.

Personally, i believe in sending it the way I proposed mostly because it conveys more information. If you see download in regular intervals (10, 20, 30) it means that the download is likely slow. If you see skips in the download (20 -> 50) that means that your bundle is really small and that it gets downloaded quickly

**Comment by riderx** [comment]
@WcaleNieWolny i’m not sure to understand what you mean by it doesn’t work since long i see the events everyday in many users, i helped a customer yesterday with that.
I want this PR to keep the exact same behavior. 
one task one PR :)

**Comment by WcaleNieWolny** [comment]
@riderx here is the issue explained in details

https://github.com/user-attachments/assets/ef8757c1-d5eb-4084-b124-daa1f95e3901

**Comment by riderx** [comment]
Current system do send only for ten X percent so we want the same. 
stats have to be same between os and version changing that is not planned. 
so implementation of @YanisOUALAN for me is correct

**Comment by riderx** [comment]
Hey @YanisOUALAN do you confirm your proposal was implemented?

**Comment by riderx** [comment]
If not can you please do a quick PR to fix it, i didn’t saw it was missing

**Comment by YanisOUALAN** [comment]
> Hey @YanisOUALAN do you confirm your proposal was implemented?

Hello, it wasn't implemented when this PR was merged since I didn't had confirmation if the behavior i suggested was the one needed, i am going to make another commit to implement my proposal later this day or tomorrow 👍

**Comment by riderx** [comment]
Thanks 🙏 
I think we got this issue today maybe related
https://github.com/Cap-go/capgo/issues/804

**Comment by riderx** [comment]
The issue reported is in android, did you had it working on android ?

**Comment by riderx** [comment]
I merged your fix in ios

**Comment by YanisOUALAN** [comment]
> The issue reported is in android, did you had it working on android ?

It did work, but the user reported that he used the 5.9.5 version, so I doubt that it's related to this PR

**Comment by riderx** [comment]
ok great thanks i will check with the user

**Comment by riderx** [comment]
@YanisOUALAN I think there is something odd: https://github.com/Cap-go/capgo/issues/804
Why do signature is enforced? This is supposed to be only enforced if present in the config

**Comment by riderx** [comment]
@YanisOUALAN i think i found out and fixed it, it came from merge i think

**Comment by YanisOUALAN** [comment]
> @YanisOUALAN i think i found out and fixed it, it came from merge i think

Oh okay, i applied a fix in #437, but if it has been resolved, then we can just close it i guess
